Create and reuse a single etcd client

enj opened the pull request:
TODO

Signed-off-by: Monis Khan <mok@microsoft.com>

/kind bug

Fixes #111622

```release-note
TODO
```

enj committed:
wip001

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip002

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip003

Signed-off-by: Monis Khan <mok@microsoft.com>

negz commented:
Cross-linking https://github.com/kubernetes/kubernetes/issues/111622, which I believe this PR is intended to fix.

alexzielenski commented:
/triage accepted
/cc @jpbetz in case you have knowledge of this

enj commented:
This does not currently handle the `etcdLocation` override support:

https://github.com/kubernetes/kubernetes/blob/233f9c210a995b2ec8b680f25051ea656e8172db/staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go#L118-L122

We also may need a loop during client construction:

https://github.com/kubernetes/kubernetes/blob/233f9c210a995b2ec8b680f25051ea656e8172db/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L151-L165

enj committed:
wip004

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip005

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip006

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip007

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip008

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip009

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip010

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip011

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip012

Signed-off-by: Monis Khan <mok@microsoft.com>

enj commented:
> This does not currently handle the `etcdLocation` override support:
> 
> https://github.com/kubernetes/kubernetes/blob/233f9c210a995b2ec8b680f25051ea656e8172db/staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go#L118-L122
> 
> We also may need a loop during client construction:
> 
> https://github.com/kubernetes/kubernetes/blob/233f9c210a995b2ec8b680f25051ea656e8172db/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L151-L165

Took care of these as well.

enj commented:
Still need to write an integration test but I confirmed locally that this makes a default API server start create a single connection to etcd instead of ~60.  Creating a lot of CRDs does not seem to lead to new connections (maybe that got fixed in some of the earlier refactors?).

/retest

enj commented:
> Creating a lot of CRDs does not seem to lead to new connections (maybe that got fixed in some of the earlier refactors?)

Nevermind - once I fetched the CRs I was able to see the extra connections against `master`.  With this change there is only a single connection to etcd, even with 100s of CRDs.

enj commented:
/retest

fedebongio commented:
/cc @serathius

fedebongio commented:
/cc @logicalhan @wenjiaswe

stevekuznetsov commented:
/assign

stevekuznetsov commented:
/assign @jpbetz

jpbetz commented:
/assign

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -126,3 +151,182 @@ func NewDefaultConfig(prefix string, codec runtime.Codec) *Config {
+
+func (c *TransportConfig) Complete(ctx context.Context) error {
+	if c.complete {
+		return fmt.Errorf("TransportConfig.Complete called more than once")
```
Not sure if this would be acceptable or not, but if this is going to be a process-ending error anyway (right?), could panic here and not require the error propagation changes up the stack.

negz commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -126,3 +151,182 @@ func NewDefaultConfig(prefix string, codec runtime.Codec) *Config {
+		dialOptions = append(dialOptions, grpc.WithContextDialer(dialer))
+	}
+
+	l, err := logutil.CreateDefaultZapLogger(etcdClientDebugLevel())
+	if err != nil {
+		return nil, err
+	}
+	etcd3ClientLogger := l.Named("etcd-client")
```
This was introduced in https://github.com/kubernetes/kubernetes/pull/111477, as a workaround for the excessive memory consumed by creating many etcd client loggers. Given that this PR causes Kubernetes to create only one per transport it should be safe to return to letting etcd use its default client (by not passing in `Logger` below).

enj replied:
Ah nice, let me do that.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -320,7 +174,7 @@ func startCompactorOnce(c storagebackend.TransportConfig, interval time.Duration
 	defer compactorsMu.Unlock()
 
-	key := fmt.Sprintf("%v", c) // gives: {[server1 server2] keyFile certFile caFile}
+	key := fmt.Sprintf("%v", c) // gives: {[server1 server2] keyFile certFile caFile}  // TODO does this make sense?
```
Can we just use `c` as the key now that we're creating one `TransportConfig` per override?

enj replied:
Reading `startCompactorOnce` and `etcd3.StartCompactor` makes me think this whole thing is subtly broken today (the interval checking and compactor replacement does not line up with how duplicate endpoints are handled and creating a string key from a struct with functions has dangerous edges).  Not sure what the best way to fix this is.

jpbetz replied:
FWIW, I would feel a lot more comfortable if we just had something like:

```go
type TransportConfigKey struct {
	ServerList    string // strings.join() of the TransportConfig.ServerList
	KeyFile       string
	CertFile      string
	TrustedCAFile string
}
```

Then at least there is not question what the key is.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -52,6 +74,9 @@ type TransportConfig struct {
 	// The TracerProvider can add tracing the connection
 	TracerProvider oteltrace.TracerProvider
+
+	complete  bool
+	getClient func() (*clientv3.Client, error)
```
It wasn't until I got familiar with this code that the `complete` field, and how it guards `getClient` made sense to me.  Maybe add a small comment explaining?  An alternative would be to omit `complete` and check if `getClient == nil` to know if the config is complete, but I don't know if others would find that easier to follow.

enj replied:
I will work on adding comments through the diff to make it easier to follow.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -390,7 +236,6 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc func() runtime.
 		once.Do(func() {
 			stopCompactor()
 			stopDBSizeMonitor()
-			client.Close()
```
Is this 100% safe? For the kube-apiserver I can't think of any actual problems, but this staging repo allows anyone to write a binary that starts an API server, this appears to leak a client in that case..

enj replied:
This `client` always comes from `newETCD3Client` which always comes from `TransportConfig.Client` which is always lifecycled with the `ctx` passed into `TransportConfig.Complete`.  Since the fields are private, it is impossible to get a client without calling `Complete` meaning any custom code that calls into this function would be at fault if it chose to pass in something like `context.TODO()`.  The standard wiring used by aggregated API servers always passes in a valid context.  Is there some other leak that you are seeing?

jpbetz replied:
Ignore me, I had missed the `close()` in `getClient()` :stuck_out_tongue:

enj committed:
wip013

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip014

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip015

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip016

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip017

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip018

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip019

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip020

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip021

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip022

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip023

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip024

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
wip025

Signed-off-by: Monis Khan <mok@microsoft.com>

enj commented:
@jpbetz @stevekuznetsov PTAL - the only thing left to do is clean up the commit message and squash.  I have addressed all comments and added integration tests.  I confirmed that the integration tests fail against `master` but pass against this PR.

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -126,3 +148,184 @@ func NewDefaultConfig(prefix string, codec runtime.Codec) *Config {
+	// decorate the KV instance so we can track etcd latency per request.
+	client.KV = etcd3.NewETCDLatencyTracker(client.KV)
+
+	// lifecycle the etcd client via ctx
```
```suggestion
	// bind the lifecycle of the etcd client to this context
```

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -304,10 +160,28 @@ type runningCompactor struct {
 	refs     int
 }
 
+// transportConfigKey is the subset of storagebackend.TransportConfig fields used for equality comparisons.
```
```suggestion
// transportConfigKey is the subset of storagebackend.TransportConfig fields used for equality comparisons when spinning up compaction goroutines.
```

bart0sh commented:
/king bug

wojtek-t commented:
/assign

bart0sh commented:
@enj  If you still need this PR then please rebase, if not, please close the PR.

aojea commented:
@mborsz @wojtek-t based on the observation that one busy object can block an entire connection, maybe is safer to use independent clients per objects and not proceed with this

enj commented:
> @mborsz @wojtek-t based on the observation that one busy object can block an entire connection, maybe is safer to use independent clients per objects and not proceed with this

Uh where did we observe this?  If that is the case, that sounds like a bug in the etcd client.  The general approach we do with http clients is to reuse the client/TLS config to get maximum connection reuse.  How is this different?

aojea commented:
> Uh where did we observe this? If that is the case, that sounds like a bug in the etcd client. The general approach we do with http clients is to reuse the client/TLS config to get maximum connection reuse. How is this different?

will try to reproduce later and share here

wojtek-t commented:
@enj - sorry for the delay from our side

The problem that @aojea mentioned is now summarized here: https://github.com/etcd-io/etcd/issues/15402 and its not an artificial one - we faced that in one of our production clusters.

I agree that this is a bug, but that makes me more afraid about sharing connections across resources - as exposes us further towards the risk that some CRD load may starve our critical resources...
Even when we fix this bug, we can't be sure there aren't any other similar things.

I also have a temptation to make this happen, but given that I don't think having multiple etcd clients is causing us significant pain, I'm more on the side of maintaining status-quo here...

enj commented:
@wojtek-t nice find.  I guess Jordan's concerns were warranted 😁

Since we seem to be okay with the resource utilization caused by having multiple etcd clients and it appears to be accidentally protecting us from a class of etcd bugs, I agree that we should leave this alone for now.  We can always revisit it later if we decide that a single shared client would provide us some benefit.

wojtek-t commented:
@enj - thanks for understanding and sorry for wasting your time - I guess my opinion about it also evolved over time...

wojtek-t commented:
@enj - Actually, I was just chatting with Jordan about it and he came up with a nice idea that instead of reusing etcd client across everything, maybe we should consider reusing etcd client on across all CRDs (since they are theoretically unbounded) and leave built-in resources as is (with etcd-client per built-in resource).

While it still has some risk inside, I think it's a nice proposal that eliminates the inherent risk that it will break the built-in functionality. So I would be supportive for it (and Jordan would be too as it was his idea :) )

It may require non-trival changes to this PR, but maybe we should consider it?

stevekuznetsov commented:
If possible, it would be awesome to see an option for the size of the client pool to be shared across CRDs. Certain installations (think CrossPlane with AWS provider) might have O(1k) CRDs and would likely want to tune this sort of thing. cc @hasheddan

bart0sh commented:
@enj please fix CI failures, set kind and rebase, thanks.
